STYLING.md

```diff
@@ -0,0 +1,225 @@
+# React Native Styling Guidelines
+
+## Where to Define Styles
+
+All styles must be defined in the `styles.js` file which exists as a globally exported stylesheet of sorts for the application. Unlike some React Native applications we are not using `StyleSheet.create()` and instead store styles as plain JS objects. There are many helper styles available for direct use in components and can be found in the `/styles` directory.
+
+These helper styles are loosely based on the [Bootstrap system of CSS utility helper classes](https://getbootstrap.com/docs/5.0/utilities/spacing/) and are typically incremented by units of 4.
+
+**Note:** Not all helpers from Bootstrap exist, so it may be necessary to create the helper style we need.
+
+## When to Create a New Style
+
+If we need some minimal set of styling rules applied to a component then it's almost always better to use an array of helper styles rather than create an entirely new style if it will only be used once. Resist the urge to create a new style for any new element added to a screen. There is a very good chance the style we are adding is a "single-use" style.
```

Comment:
I feel like this is _mostly_ correct based on a lot of what I'm seeing so far. But maybe @shawnborton has some thoughts on this.

Another thought I had was that maybe it's OK to move an array of styles to `styles.something` just from a code organization perspective. Since otherwise we'd have these massive lists inline arrays of helpers everywhere (which is kind of where we're headed now).

Still, it's difficult at times to tell whether to create a new style in `styles.js` or to use a big array of helpers.

Reply:
I personally prefer to see a new style created whenever we're composing more than two or three styles inline (especially if none of those styles are conditional). My reasoning is that we don't want to muddle our JSX with too much style logic inline, because I think it reduces readability of the code. And composing 4 new styles is basically equivalent to creating a new one. Even in `style.js`, however, we can take advantage of all the utilities such as our color themes, spacing, fonts, flex, etc... that are nice to be able to adjust globally.

Maybe we could provide some concrete guidelines by imposing a rule-of-three here and agree concretely that composing three styles inline is acceptable, but more than that should be created in a separate file.

Reply:
Thanks. I _think_ this is probably the most contentious thing I want to introduce here - but mostly providing this as a starting point. I think the suggestion to have no more than inline 3 helpers is interesting, but I'm not totally confident about it. I worry it can lead to lots of "one off" styles that are never re-used again. But maybe that's OK?

Would love to get @shawnborton's take on this if he's got some perspective to share that might help us settle on something. I'm not 100% sure what would be best and would like to lean on his experience to help establish a baseline.

Reply:
I think if the component will be a truly reusable component, it makes sense to try to put all of the styles together under one "class." Like take buttons for example, we don't want someone to need to know to use the right border radius, background color, font color, and font size. So I actually think I lean a little bit towards the opposite of what you are saying. 

Now I do think we should encourage people to use helper styles for things like spacing, layout, flex, etc. But any kind of component that will be reused multiple times, let's give it a class. But maybe one thing to distinguish is that each reusable component is essentially made up of other smaller components. So you should always use (existing) variables for things like font size, colors, radius, etc. Thoughts on that?

Reply:
Sorry for the late reply here @shawnborton. 

>we should encourage people to use helper styles for things like spacing, layout, flex, etc. But any kind of component that will be reused multiple times, let's give it a class.

I think you are essentially saying we should do a "class" or style definition over a React component, but want to clarify since "component" could mean different things to different people.

If we could distill what you are saying into some kind of rule what would it be?

>I lean a little bit towards the opposite of what you are saying

It sounds like we are generally agreeing that "single use styles" are bad but not sure which thing you are opposed to, so let us know.

Reply:
Yeah, maybe we are actually agreeing on most of this. I agree that single use styles are bad, we want people to look at the existing design system and reuse what's there if possible rather than creating something new for a one-off scenario.

> I think you are essentially saying we should do a "class" or style definition over a React component, but want to clarify since "component" could mean different things to different people.

Totally, from a non-technical perspective, I am saying that that any kind of UI object (button, form, chat row, etc) that will be reused, let's encapsulate it into one object that has additional modifiers. Let me know if that generally makes sense.

Reply:
Nice. So, by encapsulate into one object that has additional modifiers do we just mean adopt some naming convention where the modifiers are also "classes"? Probably the `button*` styles are the clearest example of where we've begun to do something like this...

```js
const styles = {
    button: {},
    buttonSuccess: {},
    buttonDisable: {},
    buttonSmallText: {},
}
```

Reply:
Yup I think that's a great example. Though in the future, I think the button component would be a great candidate for some kind of prop/value convention where we'd have something roughly like `<button size=small style=secondary styles={anything extra could go here like width utilities, etc}>`

Reply:
Cool, yeah I agree with the prop/value convention. Part of my motivation with adding the `STYLING.md` stuff is to maybe codify these ideas into a set of rules so we have an easy way to guide people on how to approach styles and refactorings...

What if we did something like:

- If an array of styles is used 3 times it should be given it's own style object e.g. `styles.button`
- Variations on that style should follow a naming convention of `styles.componentModifer` e.g. `styles.buttonSuccess`
- If a major style has more than 3 usages and also multiple modifier styles it should be moved into a re-usable component so styles can be modified via props

I think that might be a good framework to start with ?

Reply:
That sounds great to me.